source/common/access_log/access_log_formatter.cc

```diff
@@ -64,14 +64,67 @@ std::string FormatterImpl::format(const Http::HeaderMap& request_headers,
   std::string log_line;
   log_line.reserve(256);
 
-  for (const FormatterPtr& formatter : formatters_) {
-    log_line +=
-        formatter->format(request_headers, response_headers, response_trailers, stream_info);
+  for (const FormatterProviderPtr& provider : providers_) {
+    log_line += provider->format(request_headers, response_headers, response_trailers, stream_info);
   }
 
   return log_line;
 }
 
+JsonFormatterImpl::JsonFormatterImpl(
+    const std::map<const std::string, const std::string>& format_mapping) {
+  for (const auto& pair : format_mapping) {
+    auto providers = AccessLogFormatParser::parse(pair.second);
+
+    // Enforce that each key only has one format specifier in it
+    if (providers.size() > 1) {
+      throw EnvoyException(fmt::format(
+          "More than one format specifier was provided in the JSON log format: {}", pair.second));
+    }
+
+    if (providers.size() < 1) {
+      throw EnvoyException(
+          fmt::format("No format specifier was provided in the JSON log format: {}", pair.second));
+    }
+
+    json_output_format_.emplace(pair.first, std::move(providers[0]));
+  }
+}
+
+std::string JsonFormatterImpl::format(const Http::HeaderMap& request_headers,
+                                      const Http::HeaderMap& response_headers,
+                                      const Http::HeaderMap& response_trailers,
+                                      const StreamInfo::StreamInfo& stream_info) const {
+  auto output_map = this->toMap(request_headers, response_headers, response_trailers, stream_info);
+
+  ProtobufWkt::Struct output_struct;
+  for (const auto& pair : output_map) {
+    ProtobufWkt::Value string_value;
+    string_value.set_string_value(pair.second);
+    (*output_struct.mutable_fields())[pair.first] = string_value;
+  }
+
+  std::string log_line;
+  auto conversion_status = ProtobufUtil::MessageToJsonString(output_struct, &log_line);
+  if (!conversion_status.ok()) {
+    throw EnvoyException(
```

Comment:
We can't safely throw exceptions at runtime on the data path in Envoy, only during construction (config ingestion). Otherwise, Envoy will behave badly. We should have a more graceful failure in this scenario.

Reply:
Cool, which do you prefer:
- Return an empty string
- Return a string containing something to the effect of `Error converting access log output to JSON`
- pointer shenanigans to return an error code (I'm against this personally)

Reply:
Probably JSON conversion error, it will be easier to understand WTF happened when this fails.

Reply:
Done